### DISCUSSION TITLE
Optional Parameters in any order

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#2622] [Language Ideas] [answered]
Right now I write code like the following:
````
public void Foo(IComparer<T> MyComparer = default, FooObject Foo = default, AnotherObject Another = default){
  MyComparer ??= Comparer<T>.Default;
  Foo ??= FooObject.Default;
  Another ??= AnotherObject.Default;
}
````

and I call it like follows:
````
Foo(MyComparer); //This is nice.
Foo(default, new FooObject()); //Not as much.
Foo(Another: new AnotherObject); //Don't like at all.
````

It would be really nice if C# could infer the position of optional parameters by their type so I could simply say:
````
Foo(MyComparer());
Foo(new FooObject());
Foo(new AnotherObject());

````

**Comment by HaloFour** [comment, marked as answer]
Considering that methods can have multiple optional parameters of the same type I don't see how that could work.  It also sounds like an unmitigated nightmare where it affects overload resolution.

The C# compiler does now allow you to specify optional parameters in any order by name, so the third example: `Foo(Another: new AnotherObject());` does work.  Is that not sufficient?
